### PR TITLE
Adds BaseItem.Equals method that compares items by their Id

### PR DESCRIPTION
**Pull request by JustAMan**
**Changes**

Speed up equality comparison of BaseItem. Should help in `.Distinct()` calls and so on.

**Issues**

N/A

### CODE REVIEW

**Comment by mark-monteiro** [comment]
An override for `GetHashCode` should probably also be added here. You should be able to generate the code for this in Visual Studio: https://docs.microsoft.com/en-us/visualstudio/ide/reference/generate-equals-gethashcode-methods?view=vs-2019

**Comment by JustAMan** [comment]
Thanks for suggestion. I don't use VS in JF development for now (sticking to VSCode), so I made a manual implementation.

**Comment by mark-monteiro** [comment]
@Bond-009 You might be able to provide some more insight here, but I worry that this change could introduce some subtle bugs if there is any code that relies on the existing behavior of reference equality comparison (or maybe instead it would fix some subtle bugs :p). But it would be difficult to check for that manually.

**Comment by JustAMan** [comment]
Well, `BaseItem.Id` is a GUID which is a key column in the SQL schema, so I highly doubt we're going to have different items with equal ids. If we do have some that is really a bug :)

**Comment by mark-monteiro** [comment]
@JustAMan Some cases where behavior is altered:

```
var list1 = fetchsomeitems();
var list2 = fetchsomeitems();
var set = list1.concat(list2).ToHashSet();

var item1 = fetchanitem();
var item2 = fetchthesameitem();
var areEqual = item1.equals(item2);
```

Before this change `set` would always contain all elements from both lists, but after the change items with duplicate ids would not be included.
Before this change `areEqual` would be `false` but after this change it would be `true`.
You'd also have issues when using `BaseItem` as a dictionary key, etc.

These are subtle differences, but because `BaseItem` is such a commonly used item it would be really difficult to check the code to see if there are any cases where this behavior change will have a negative impact. I agree that this is probably the behavior we want, but maybe this should wait until after release to avoid introducing bugs?

**Comment by JustAMan** [comment]
This certainly **shouldn't be merged in 10.5.0** (note that it is not added to the project).

Basically, the only case where this thing could make a negative impact that I can think of is if we have different media items (with different types) but matching ids.
Even if we have those, I would still claim that in the long run their existence is a bug.

**Comment by mark-monteiro** [comment]
> Why, though? I thought that `object.Equals()` compares all fields (dunno if it uses properties), so `areEqual` should already be `true` before the change.

The default implementation does reference equality and not check member or property values (See the `Remarks` section here: https://docs.microsoft.com/en-us/dotnet/api/system.object.equals?redirectedfrom=MSDN&view=netframework-4.8#System_Object_Equals_System_Object_):
> If the current instance is a reference type, the Equals(Object) method tests for reference equality, and a call to the Equals(Object) method is equivalent to a call to the ReferenceEquals method.

**Comment by JustAMan** [comment]
> The default implementation does reference equality and not check member or property values

In this case a `listOfItems.Distinct(otherList)` is completely useless unless `.Equals()` is overridden :)

**Comment by mark-monteiro** [comment]
> In this case a `listOfItems.Distinct(otherList)` is completely useless unless `.Equals()` is overridden :)

Yeah, I was considering implementing `IEqualityComparer` for the playlist duplicate checking instead, but I think we can just wait to merge the playlist issue in the next release as well.

I'm going to add a few more comments, but I'm convinced the overall approach here is good since it's going into the next release.

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 1]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
         public static readonly IReadOnlyCollection<ExtraType> DisplayExtraTypes = new[] { Model.Entities.ExtraType.BehindTheScenes, Model.Entities.ExtraType.Clip, Model.Entities.ExtraType.DeletedScene, Model.Entities.ExtraType.Interview, Model.Entities.ExtraType.Sample, Model.Entities.ExtraType.Scene };
 
         public virtual bool SupportsExternalTransfer => false;
+
+        public override bool Equals(object obj)
+        {
+            Logger.LogDebug("Comparing me ({0}) to generic them ({1})", this, obj);
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: No need to log here I think

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 2]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
...
+            return this.Equals(obj as BaseItem);
+        }
+
+        public bool Equals(BaseItem item)
+        {
+            Logger.LogDebug("Comparing me ({0}) to specific them ({1})", this, item);
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: No need to log here either

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 3]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
...
+            {
+                return false;
+            }
+
+            return Id == item.Id;
+        }
+
+        public override int GetHashCode()
+        {
+            return Id.GetHashCode();
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: Need to handle null values for `Id` here. Visual Studio generates this with `HashCode.Combine(Id);` which can handle the null check.

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 4]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
         public static readonly IReadOnlyCollection<ExtraType> DisplayExtraTypes = new[] { Model.Entities.ExtraType.BehindTheScenes, Model.Entities.ExtraType.Clip, Model.Entities.ExtraType.DeletedScene, Model.Entities.ExtraType.Interview, Model.Entities.ExtraType.Sample, Model.Entities.ExtraType.Scene };
 
         public virtual bool SupportsExternalTransfer => false;
+
+        public override bool Equals(object obj)
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: Add <inheritsdoc/> to these functions.

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 5]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
...
+
+        public bool Equals(BaseItem item)
+        {
+            Logger.LogDebug("Comparing me ({0}) to specific them ({1})", this, item);
+            if (item == null)
+            {
+                return false;
+            }
+
+            return Id == item.Id;
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: Need to check for null `Id` here. `object.Equals(Id, item?.Id)` should do the trick I think (and then you can get rid of the null check above as well)

File: MediaBrowser.Controller/Entities/BaseItem.cs [hunk 6]
```diff
@@ -2918,5 +2918,27 @@ public virtual long GetRunTimeTicksForPlayState()
         public static readonly IReadOnlyCollection<ExtraType> DisplayExtraTypes = new[] { Model.Entities.ExtraType.BehindTheScenes, Model.Entities.ExtraType.Clip, Model.Entities.ExtraType.DeletedScene, Model.Entities.ExtraType.Interview, Model.Entities.ExtraType.Sample, Model.Entities.ExtraType.Scene };
 
         public virtual bool SupportsExternalTransfer => false;
+
+        public override bool Equals(object obj)
+        {
+            Logger.LogDebug("Comparing me ({0}) to generic them ({1})", this, obj);
+            return this.Equals(obj as BaseItem);
```
Comment (mark-monteiro) [inline MediaBrowser.Controller/Entities/BaseItem.cs]: Not 100% sure, but you might get marginally better performance doing `return obj is BaseItem baseItem && Equals(baseItem);` Personally I find that a bit more readable as well.

Comment (Bond-009) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: Only diff in perf would be 1 less function call. Doing `is` or using `as` and checking for `null` is practically the same

File: MediaBrowser.Controller/Entities/BaseItem.cs (hunk 4 above)
Comment (JustAMan) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: Add what?

File: MediaBrowser.Controller/Entities/BaseItem.cs (hunk 5 above)
Comment (JustAMan) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: neat, thx :)

File: MediaBrowser.Controller/Entities/BaseItem.cs (hunk 6 above)
Comment (JustAMan) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: I don't believe in much perf difference here (and especially I don't believe it matters a lot in this place), but it certainly is more readable. Thanks!

File: MediaBrowser.Controller/Entities/BaseItem.cs (hunk 4 above)
Comment (Bond-009) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: Maybe `/// <inheritdoc />`?

Comment (JustAMan) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: Added

Comment (mark-monteiro) [reply MediaBrowser.Controller/Entities/BaseItem.cs]: Sorry, yes meant to write `/// <inheritdoc/>`

**Comment by JustAMan** [comment]
Marking as `blocked` so as to _not_ accidentally merge in upcoming 10.5.0

**Comment by mark-monteiro** [comment]
Unblocking now that 10.5.0 has been released.

**Pull request merged**